:sparkles: Feat: 스크랩/카테고리 삭제 기능

JJongsKim opened the pull request:
## 📮 관련 이슈

## ⛳️ 작업 내용

스크랩페이지의 카테고리들을 선택 후 삭제할 수 있는 기능을 제작했습니다.
- 카테고리 별, 카테고리 디테일, 콘텐츠 별 selected[ref.current] 값에 따라 요청하는 API가 다르도록 분리
- 스크랩 삭제 부분은 카테고리 디테일 목록과 콘텐츠 목록이 새로고침 될 수 있도록 쿼리키 추가
- selected 초기화 함수는 API 요청 성공 시 일어나도록 코드 이동

## 🌱 PR 포인트

## 📸 스크린샷
|스크린샷|
|:--:|
|![ezgif com-gif-maker](https://user-images.githubusercontent.com/81777778/213734825-acd55bea-8252-4be9-be0b-c221f77863e9.gif)|

## 📎 레퍼런스

<!-- 참고한 레퍼런스가 있다면 기록해주세요 -->

JJongsKim committed:
:sparkles: Feat: 스크랩 카테고리 삭제 구현중

- 이벤트핸들러나 useEffect를 활용해도 바로 업데이트가 안되는 이슈

JJongsKim committed:
:recycle: Refactor: 이벤트핸들러로 삭제 기능 구현하도록 코드 수정중

- useEffect를 사용한 삭제코드 우선 삭제

JJongsKim committed:
:construction: Feat: 임시 삭제 기능

- useEffect 사용, 하위 컴포넌트 업데이트가 되지 않아 새로고침해야 삭제 반영되어있음

JJongsKim committed:
:sparkles: Feat: 카테고리 삭제 후 목록 업데이트 반영

- useQueryClient, invalidateQueries 추가하여 getCategories 키에 바로 반영될 수 있도록 수정

JJongsKim committed:
:sparkles: Feat: 스크랩 삭제 API hook 작성, DeleteScrapRequest 타입 변경

- 스크랩 삭제 파라미터를 리스트로 받을 수 있도록 수정
- 스크랩 삭제 API hook을 만들어 카테고리, 콘텐츠 별로 업데이트 되도록 queryKey로 추가

JJongsKim committed:
:construction: Feat: 스크랩 삭제 기능 진행중

- selected[ref.current] 값에 따라 id가 담기도록 props 설정

JJongsKim committed:
:sparkles: Feat: 스크랩 삭제 진행중

JJongsKim committed:
Merge branch 'dev' into feat/scrap-category-delete

JJongsKim committed:
:sparkles: Feat: 스크랩 삭제 기능 구현(카테고리 별 디테일 스크랩, 콘텐츠 별 스크랩)

- 스크랩 삭제 시 업데이트 적용될 수 있도록 쿼리를 수정
- selected.ref 값에 따라 요청하는 API가 다르도록 구분
- 요청 성공 시 스크랩 페이지 초기화

JJongsKim committed:
:bug: Fix: 삭제된 Id 중복되는 문제 해결

elbica commented on src/application/store/scrap/categoryState.ts:
```diff
@@ -0,0 +1,22 @@
+  default: [],
+});
+
+export const categoryDeleteList = selector({
+  key: 'categoryDeleteList',
+  get: ({ get }) => get(categoryIdsArray),
+});
```
따로 값 가공을 하지않아서 `const useCategoryDeleteList = useRecoilValue(categoryIdsArray)` 이랑 똑같은 동작을 할 것 같은데, selector을 따로 만드신 이유가 있나요?

그리고 category 관련 상태인데 `scarp/` 폴더에 있어서 헷갈리는 것 같아요
`store/category`로 분리하는 건 어떨까요?

JJongsKim replied:
아 원래는 set 변환을 하고 싶어서 selector로 선언을 한 거였는데, 어차피 atom만 사용하는거라 굳이 필요없겠네욥.
recoil이랑 폴더구조 모두 반영해서 수정하겠습니다~~~

elbica commented on src/containers/scrap/PhotoListContainer.tsx:
```diff
@@ -36,16 +51,17 @@ const PhotoListContainer = ({ data, select, onEndReached }: PhotoListContainerPr
-              src={getSrcByType(photo)}
-              text={photo.content}
-            />
+            <div key={photo.id} onClick={() => selectItem && selectCategoryDetail(photo.id)}>
+              <Photo
+                custom={css`
+                  aspect-ratio: 1/1;
+                `}
+                onClick={() => !select && router.push(`/scrap/${photo.id}`)}
```
`<Photo />`의 `onClick` 핸들러와 합치는게 좋아보여요!
둘다 클릭 했을 때 일어나는 로직인데 흩어져 있으니 파악하기가 어렵네요..!!
`handleClickPhoto` 핸들러로 합쳐서 `<Photo /> onClick` 핸들러로 내려주는 것이 어떨까요?

JJongsKim replied:
그렇게 정리하는 게 더 한 눈에 알아보기 쉬울 것 같네용! 
반영하겠습니당 :)

elbica commented on src/pages/scrap/index.tsx:
```diff
@@ -107,7 +172,11 @@ const Scrap: NextPage = () => {
           >
-            {selected[ref.current] ? '취소' : <Image src={'/icon/multiSelect.svg'} width={18} height={18} />}
+            {selected[ref.current] ? (
+              <p onClick={() => resetCategoryList()}>취소</p>
```
`onClick={resetCategoryList}`로 간략화 하는건 어떨까요?

elbica commented on src/pages/scrap/index.tsx:
```diff
@@ -66,6 +80,57 @@ const Scrap: NextPage = () => {
+    isDeleteState(!deleteState);
+  };
+
+  useEffect(() => {
+    if (deleteState && selected.category) {
+      requestDeleteCategory();
+    }
+    if (deleteState && (selected.categoryInfo || selected.content)) {
+      requestDeleteScrap();
+    }
+    isDeleteState(false);
+  }, [deleteState, isDeleteState, selected, requestDeleteCategory, requestDeleteScrap]);
```
이벤트 핸들러로 바로 삭제시키는 것이 아니라 `useEffect`로 삭제 상태가 변경되었을 때 삭제하는 식이라서, 의존성 배열도 복잡해지고 `useCallback`으로 함수들을 계속 감싸주어야 해서 가독성이 조금 떨어지는 것 같아요!
그런데 이벤트 핸들러 방식으로 하면 id들이 제대로 안내려오는 버그가 있었다고 하신거 맞죠?? (정녕 이 방법밖에 없는 것인지..)

elbica replied:
```tsx
  const handleDeleteScrap = () => {
    ...
    popup(DeletePopup, 'success');
  };
  const showDeleteScrapToast = () => show({ content: <DeleteScrapToast onDelete={handleDeleteScrap} /> })
  
  setNavigation(selected[key] ? <DeleteNavigation onClick={showDeleteScrapToast} /> : 'default');
```
디버깅 좀 해봤는데 원인은 `setNavigation()` 였습니다
삭제 모드로 들어갈때 바텀 네비게이션이 `<DeleteNavgiation />`으로 바뀌는데, 여기에 전달되는 핸들러들이 갱신이 안돼서 새로운 recoil 상태를 바라보지 못하네요
- 삭제 카테고리를 선택할 때 `page/scrap/index.ts`는 리렌더링이 일어나는데
- 이전에 이미 set 되어버린 `BottomNavigation`은 상위 hoc`withNavigation`에 있기 때문에 `page/scrap/index.ts` 리렌더링 영향을 받지 않아요 
- 리렌더링에 영향을 받지 않는다는 것은, 이전에 `<BottomNavigation />`이 처음에 set 될때의 상태인 `categoryDeleteItem`가 아예 비워져 있을때만 바라본다는 것을 뜻합니다
- 한마디로 `<DeleteNavigation />`만 이전 렌더링 상태를 바라보고, `page/scrap/index`는 리렌더링 된 갱신된 상테를 바라보는 것이 원인이었습니다 ! (* 각 리렌더링 마다 리액트 내부의 상태는 클로저로 격리되어서 관리됩니다)
- 네비게이션 설계를 잘못했네요..ㅎㅎ😖


### 해결 방법
`<BottomNavigation />`을 hoc로 감싸지 말고 `page/scrap/index` 아래에 두어 리렌더링에 영향받게 하면 됩니다
또는 `withNavigion` hoc의 `BottomNavigationContext`에 클릭 핸들러를 넣어서 리렌더링마다 계속 갱신하는 방법도 있겠네요

```tsx
const handleDeleteScrap = () => {
    if (selected.category) {
      categoryMutation.mutate({ ids: categoryDeleteItem }, { onSuccess: resetCategoryMutation });
    }
    if (selected.categoryInfo || selected.content) {
      scrapMutation.mutate({ ids: scrapDeleteItem }, { onSuccess: resetScrapMutation });
    }

    resetCategoryStates();
    popup(DeletePopup, 'success');
  };

  const showDeleteScrapToast = () => show({ content: <DeleteScrapToast onDelete={handleDeleteScrap} /> });
 
  return (
      ...
      {selected[ref.current] ? <DeleteNavigation onClick={showDeleteScrapToast} /> : <BottomNavigation />}
  )

...

export default withAuth(Scrap);

```

JJongsKim replied:
흐어어어어엉 값 갱신 이슈를 이렇게 발견하시다니,, ! 
어쩐지 매거진 페이지에선 따로 setNavigation 없이 하단 네비게이션이 나오도록 코드를 짰더니 거기는 문제 없이 이벤트핸들러로 구현이 되더라구요. 대충 네비게이션 쪽에서 값 전달이 안되는 거라고는 예상했는데 어떻게 고쳐야할지 전혀 감이 안 왔었는데,, 이렇게 이해 잘 되게 설명해주셔서 감사합니다!! 참고해서 수정해볼게요😍

elbica commented on src/containers/scrap/CategoryListContainer.tsx:
```diff
@@ -32,13 +46,16 @@ const CategoryListContainer = ({ select, onClickItem }: CategoryListContainerPro
+                onClick={() => onClickItem({ id: category.id, name: category.name })}
+                select={select}
+                // helper.ts 에러로 인해 잠시 주석화
+                // src={getSrcByType(category) ?? '/icon/scrap/defaultCategory.svg'}
```
이거는 `category` 객체 데이터를 좀 디버깅 해봐야겠네요🥲

elbica replied:
<img width="191" alt="image" src="https://user-images.githubusercontent.com/74011724/213862296-976c3414-3fd3-4117-b2f9-b2e10612328a.png">

```tsx
export const getSrcByType = (content: Scrap | Category) => {
  if (content.scrap_type === null) return '';
```
카테고리 내의 스크랩 콘텐츠가 하나도 없을 때 내부 프로퍼티들이 null로 내려오는데, `getSrcByType`에 null 체크 방어코드를 추가하는게 좋을 것 같네요!

elbica reviewed (commented):
코드가 좀 얽혀 있어서(..) 작업하기 번거로우셨을 것 같은데 잘 해주셔서 너무 수고 많으셨습니다😊
코멘트 확인 부탁드려요!

JJongsKim committed:
:recycle: Refactor: 코드리뷰 수정사항 반영(함수 간략화, 폴더구조 변경)

- onClick 이벤트 코드 작성을 간략화
- scrap과 category state, hook 함수를 서로 다른 폴더로 구분
- 사용하지 않는 selector 삭제

JJongsKim committed:
:recycle: Refactor: 코드리뷰 수정사항 반영(이벤트핸들러 수정, 주석코드 제거)

- 코드리뷰 참고하여 삭제네비게이션 scrap/index로 이동, 값 갱신 확인
- helper.ts 오류 주석코드 제거

JJongsKim committed:
:recycle: Refactor: 매거진 메인페이지 코드 리팩토링

- 삭제기능 이벤트핸들러로 수정
- 삭제 관련 매거진 state, hook 수정

JJongsKim committed:
:memo: Update: reset 네이밍 통일

JJongsKim committed:
:memo: Update: onSuccess 선언 함수 코드 변경

elbica reviewed (approved):
고생 많으셨습니다🙂
